common/src/main/java/io/pravega/common/util/SearchUtils.java

```diff
@@ -16,14 +16,100 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import io.pravega.common.MathHelpers;
-import java.util.AbstractMap;
-import java.util.Map;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
 import java.util.function.LongFunction;
 
 /**
  * A utility class to hold the searching/sorting functions.
  */
 public class SearchUtils {
+    
+    
+    /**
+     * Exactly the same as {@link #newtonianSearch(LongFunction, long, long, long, boolean)} but async.
+     * 
+     * @param getValue The function that returns a long given an index
+     * @param fromIdx The index from which to start the search
+     * @param toIdx The index at which to end the search
+     * @param target The value to search for
+     * @param greater If the element is not in the list true indicates that the next larger element
+     *            should be returned false indicates the next smaller element should be returned
+     * @return A future for an Entry object containing the closest index and its value
+     * @throws IllegalArgumentException if the input list is empty
+     */
+    public static CompletableFuture<Entry<Long, Long>> asyncNewtonianSearch(LongFunction<CompletableFuture<Long>> getValue,
+                                                                            long fromIdx, long toIdx, long target,
+                                                                            boolean greater) {
+        if (fromIdx > toIdx || fromIdx < 0) {
+            throw new IllegalArgumentException("Index size was negative");
+        } else if (fromIdx == toIdx) {
+            return getValue.apply(fromIdx).thenApply(value -> new SimpleEntry<>(fromIdx, value));
+        }
+
+        CompletableFuture<Long> fromFuture = getValue.apply(fromIdx);
+        CompletableFuture<Long> toFuture = getValue.apply(toIdx);
+
+        return fromFuture.thenCombine(toFuture, (fromValue, toValue) -> {
+            SimpleEntry<Long, Long> singleEntry = getSimpleEntry(fromIdx, toIdx, target, fromValue, toValue);
+            if (singleEntry != null) {
+                return CompletableFuture.completedFuture(singleEntry);
+            }
+
+            double beginSlope = calculateSlope(fromIdx, toIdx, fromValue, toValue);
+            double endSlope = beginSlope;
+
+            if (toIdx <= fromIdx + 1) {
+                return CompletableFuture.completedFuture(
+                    getSimpleEntryBasedOnGreater(fromIdx, toIdx, greater, fromValue, toValue));
+            }
+            long guessIdx = generateNextGuess(fromIdx, toIdx, target, fromValue, toValue, beginSlope, endSlope);
+            return recursivlyNarrowSearch(getValue, fromIdx, fromValue, toIdx, toValue, guessIdx, target, greater);
+        }).thenCompose(f -> cast(f));
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
Not sure why we need to do this.

Reply:
The signature of the public interface is a `CompletionStage<Entry<Long, Long>>` but because of bad design of the Java CompletableFuture API the result of a `thenCombine` operation will take the inferred type and add a `? extends` to the front. This is accurate, but the java type inference is not smart enough to understand that `CompletionStage<? extends Entry<Long, Long>>` by definition is a `CompletionStage<Entry<Long, Long>>` because `Entry` is an Interface and not a concrete type. So an explicit cast is required.